compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -3733,35 +3732,15 @@ public void visit(BLangAnnotAccessExpr annotAccessExpr) {
         }
     }
 
-    private boolean indirectIntersectionExists(BLangExpression expression, BType testType) {
+    private boolean intersectionExists(BLangExpression expression, BType testType) {
         BType expressionType = expression.type;
-        SymbolEnv symbolEnv = symTable.pkgEnvMap.get(env.enclPkg.symbol);
-        Types.IntersectionContext intersectionContext =
-                Types.IntersectionContext.compilerInternalIntersectionTestContext();
 
-        switch (expressionType.tag) {
-            case TypeTags.UNION:
-                if (types.getTypeForUnionTypeMembersAssignableToType((BUnionType) expressionType, testType,
-                        symbolEnv, intersectionContext) !=
-                        symTable.semanticError) {
-                    return true;
-                }
-                break;
-            case TypeTags.FINITE:
-                if (types.getTypeForFiniteTypeValuesAssignableToType((BFiniteType) expressionType, testType) !=
-                        symTable.semanticError) {
-                    return true;
-                }
-        }
+        BType intersectionType = types.getTypeIntersection(
+                Types.IntersectionContext.compilerInternalNonGenerativeIntersectionContext(),
+                expressionType, testType, env);
 
-        switch (testType.tag) {
-            case TypeTags.UNION:
-                return types.getTypeForUnionTypeMembersAssignableToType((BUnionType) testType, expressionType,
-                        symbolEnv, intersectionContext) !=
-                        symTable.semanticError;
-            case TypeTags.FINITE:
-                return types.getTypeForFiniteTypeValuesAssignableToType((BFiniteType) testType, expressionType) !=
-                        symTable.semanticError;
+        if (intersectionType != null && intersectionType != symTable.semanticError) {
```

Comment:
I don't think `getTypeIntersection` returns null

Reply:
Yeah, had the private method in mind. Fixed in 79428e3. Thanks!